django/contrib/contenttypes/fields.py

```diff
@@ -558,25 +559,37 @@ def get_queryset(self):
                 queryset = super().get_queryset()
                 return self._apply_rel_filters(queryset)
 
+        def get_content_type(self, obj):
+            return ContentType.objects.db_manager(obj._state.db).get_for_model(
+                obj, for_concrete_model=self.for_concrete_model)
+
         def get_prefetch_queryset(self, instances, queryset=None):
             if queryset is None:
                 queryset = super().get_queryset()
 
             queryset._add_hints(instance=instances[0])
             queryset = queryset.using(queryset._db or self._db)
 
+            content_type_ids = {self.get_content_type(obj).pk for obj in instances}
             query = {
-                '%s__pk' % self.content_type_field_name: self.content_type.id,
+                '%s__in' % self.content_type_field_name: content_type_ids,
```

Comment:
Current implementation of `get_prefetch_queryset()` assumes that all instances have the same content type (there is an issue), but the fix is not optimal IMO because object IDs can be the same in different content types, e.g. we have two related objects:

- object ID 1 with content type ID 1,
- object ID 2 with content type ID 2,

this query will return also:
- object ID 2 with content type ID 1,
- object ID 1 with content type ID 2,

which is not correct. I know that we are matching them below but still I think we can limit the no. of objects only to actually needed.

Reply:
I tried to change the filter with separated filters with or operator. Is it ok?